src/tagged-type.ts

```diff
@@ -70,16 +73,16 @@ class TaggedTypeEncoded {
     },
     number: (value: number): TaggedDouble | TaggedInt | TaggedLong => {
       if (`${value}`.includes(".")) {
-        return { "@double": value };
+        return { "@double": value.toString() };
       } else {
         if (value >= -(2 ** 31) && value <= 2 ** 31 - 1) {
-          return { "@int": value };
-        } else if (value >= -(2 ** 63) && value <= 2 ** 63 - 1) {
+          return { "@int": value.toString() };
+        } else if (Number.isSafeInteger(value)) {
           return {
             "@long": value.toString(),
           };
         }
-        return { "@double": value };
+        return { "@double": value.toString() };
```

Comment:
This is a bug:

```
> 2**100
1.2676506002282294e+30
```

Note also it will suffer the same precision loss as line 72.

We should be consistent.

Either:

1. reject any precision loss
2. mutate places we lose precision to doubles; and only blow up if we go past Fauna or JS's max double size.

IMO (2) is preferable. Though we need to account for precision loss with documentation or warnings to users.

Reply:
Fauna accepts the string representation with the exponent.

> mutate places we lose precision to doubles; and only blow up if we go past Fauna or JS's max double size.

I believe that we do this and that. The encoder checks for `Number.isSafeInteger`. If it is, then it's a Long. If not, then we encode as a Double.

Reply:
we don't do that for BigInts.

Reply:
example:

```
> 2**100
1.2676506002282294e+30
> 2**100 - 1
1.2676506002282294e+30
>
```

This is what i'm trying to get at with precision loss being treated the same for BigInt and number (e.g. do we allow it or blow up?)

Reply:
No, but I think that is a good idea. Once you have a BigInt, the intent is clear that it should remain a Long. You can't even do arithmatic with a BigInt without converting other values to BigInt (e.g. `LONG_MAX + BigInt(1))` just to increment). BigInt is just a stand in for a 64bit Long. It seems better to me to throw an error that the int is too big before we convert to a Double on behalf of devs.